### PR TITLE
ABILITY_SYMBIOSIS

### PR DESCRIPTION
**Pull request by AgustinGDLV**
Addresses issue #346. Lots of testing done, some of it below, a lot of it recorded and some of those recordings below, too. All major issues so far resolved! I may go back and try to combine trysymbiosis with the new MOVEEND_EFFECT which was implemented after to fix some other issues, but I'm not sure how simple that would be. _If it would be best to merge the two, let me know._

Symbiosis Showcase

**Issues:**
- ~~Gem interactions unknown, but presumably they function like Focus Sash and the like.~~
- ~~Bestow gets cut off early when Symbiosis procs.~~
- ~~If an ally eats a foe's berry with Bug Bite but it has no effect (e.g. Sitrus Berry on max HP), Symbiosis does not proc.~~
- ~~Fling is completely busted.~~
- ~~Eject Button and Eject Pack incorrectly trigger Symbiosis.~~  _Need to test in IE, wasn't working because of debug mode._
- ~~Message bug with damage-reducing berries (new item is received before message appears).~~

**Tested Interactions:**
- [x] Red Card
- [x] Focus Sash
- [x] Bug Bite (all cases!)
- [x] Knock Off
- [x] Incinerate
- [x] Air Balloon
- [x] Trick
- [x] Eject Button / Pack
- [x] Gems
- [x] Unexchangeable Items
- [x] Fling
- [x] Natural Gift
- [x] Damage-reducing Berries
- [x] Multiple Symbiosis abilities on side
- [x] Gem behavior differences in Gen 6 and Gen 7+

**Extra Demos:**
<details>
  <summary>show:</summary>

**Bug Bite:** 
Symbiosis Bug Bite (No Effect Item)
Symbiosis Bug Bite

**Focus Sash:**
Symbiosis Focus Sash
**Incinerate:**
Symbiosis Incinerate
**Knock Off:**
Symbiosis Knock Off
**Red Card:**
Symbiosis Red Card
**Gen 6 Gems:**
Symbiosis Gem + Life Orb (Gen 6)
**Gen 7 Gems:**
Symbios Gem + Life Orb (Gen 7)
**Eject Pack:**
Symbiosis Eject Pack
 
</details>

## **Discord Contact Info**
Agustin#1522

### CODE REVIEW

**Comment by AgustinGDLV** [comment]
New commit. Other than some major issues with Fling, Symbiosis is accurate and functional in an isolated BE branch. More testing required with items from IE.

Resolved Issues:
- Bestow gets cut off early when Symbiosis procs.
- If an ally eats a foe's berry with Bug Bite but it has no effect (e.g. Sitrus Berry on max HP), Symbiosis does not proc.

**Comment by AgustinGDLV** [comment]
Final commit? Tested most interactions and whatever edge cases I could think of. Gems and damage-reducing berries now function correctly (using a new MOVEEND effect), as do Eject Pack and Eject Button. Added battle config for gem function. I think this is complete!

**Comment by ghoulslash** [comment]
fix conflicts

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -7244,6 +7269,49 @@ static bool32 TryCheekPouch(u32 battlerId, u32 itemId)
...
+        gLastUsedAbility = gBattleMons[ally].ability;
+        gBattleScripting.battler = gBattlerAbility = ally;
+        gBattlerAttacker = battler;
+        BattleScriptPush(gBattlescriptCurrInstr + 2);
+        gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
+        return TRUE;
+    }
+    return FALSE;
+}
+
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:7314]:
```suggestion
static bool32 TrySymbiosis(u32 battler, u32 itemId)
{
    if (!gBattleStruct->itemStolen[gBattlerPartyIndexes[battler]].stolen
        && gBattleStruct->changedItems[battler] == ITEM_NONE
        && GetBattlerHoldEffect(battler, TRUE) != HOLD_EFFECT_EJECT_BUTTON
        && GetBattlerHoldEffect(battler, TRUE) != HOLD_EFFECT_EJECT_PACK
    #if B_SYMBIOSIS_GEMS >= GEN_7
        && !(gSpecialStatuses[battler].gemBoost)
    #endif
        && gCurrentMove != MOVE_FLING //Fling and damage-reducing berries are handled separately.
        && !gSpecialStatuses[battler].berryReduced
        && SYMBIOSIS_CHECK(battler, BATTLE_PARTNER(battler)))
    {
        BestowItem(BATTLE_PARTNER(battler), battler);
        gLastUsedAbility = gBattleMons[BATTLE_PARTNER(battler)].ability;
        gBattleScripting.battler = gBattlerAbility = BATTLE_PARTNER(battler);
        gBattlerAttacker = battler;
        BattleScriptPush(gBattlescriptCurrInstr + 2);
        gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
        return TRUE;
    }
    return FALSE;
}

```

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -9788,6 +9843,18 @@ static void Cmd_various(void)
...
+        {
+            BestowItem(gActiveBattler ^ BIT_FLANK, gActiveBattler);
+            gLastUsedAbility = gBattleMons[gActiveBattler ^ BIT_FLANK].ability;
+            gBattleScripting.battler = gBattlerAbility = gActiveBattler ^ BIT_FLANK;
+            gBattlerAttacker = gActiveBattler;
+            BattleScriptPushCursor();
+            gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
+            return;
+        }
+        break;
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:9857]:
```suggestion
    case VARIOUS_TRY_SYMBIOSIS: //called by Bestow, Fling, and Bug Bite, which don't work with Cmd_removeitem.
        if (SYMBIOSIS_CHECK(gActiveBattler, BATTLE_PARTNER(gActiveBattler)))
        {
            BestowItem(BATTLE_PARTNER(gActiveBattler), gActiveBattler);
            gLastUsedAbility = gBattleMons[BATTLE_PARTNER(gActiveBattler)].ability;
            gBattleScripting.battler = gBattlerAbility = BATTLE_PARTNER(gActiveBattler);
            gBattlerAttacker = gActiveBattler;
            BattleScriptPushCursor();
            gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
            return;
        }
        break;
```

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -5766,6 +5774,23 @@ static void Cmd_moveend(void)
...
+                    gLastUsedAbility = gBattleMons[i ^ BIT_FLANK].ability;
+                    gBattleScripting.battler = gBattlerAbility = i ^ BIT_FLANK;
+                    gBattlerAttacker = i;
+                    BattleScriptPushCursor();
+                    gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
+                    effect = TRUE;
+                }
+            }
+            gBattleScripting.moveendState++;
+            break;
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:5793]:
```suggestion
        case MOVEEND_SYMBIOSIS:
            for (i = 0; i < gBattlersCount; i++)
            {
                if ((gSpecialStatuses[i].berryReduced
                #if B_SYMBIOSIS_GEMS >= GEN_7
                    || gSpecialStatuses[i].gemBoost
                #endif
                    ) && SYMBIOSIS_CHECK(i, BATTLE_PARTNER(i)))
                {
                    BestowItem(BATTLE_PARTNER(i), i);
                    gLastUsedAbility = gBattleMons[BATTLE_PARTNER(i)].ability;
                    gBattleScripting.battler = gBattlerAbility = BATTLE_PARTNER(i);
                    gBattlerAttacker = i;
                    BattleScriptPushCursor();
                    gBattlescriptCurrInstr = BattleScript_SymbiosisActivates;
                    effect = TRUE;
                }
            }
            gBattleScripting.moveendState++;
            break;
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**